### ISSUE TITLE
autocompletion inside arbitrary strings (i.e. not in `tw` or `apply` (etc.) tagged template literals / strings)

### ISSUE THREAD

**Issue by danielweck** [tw-in-js/typescript-plugin#11]
I have tried this Visual Studio Code configuration in my projects (to no avail):

```
  "css.validate": false,
  "less.validate": false,
  "scss.validate": false,

  "editor.quickSuggestions": {
    "strings": true
  }
```

Are you able to successfully trigger the Twind VSCode plugin inside arbitrary strings?

I am asking out of curiosity. This is not at all a deal breaker for me, because I use the `tw` / `apply` tagged template string syntax anyway, in my Preact `class` attributes as well as in regular code.

I do however "erase" the `tw'...'` calls when compiling, as they add unnecessary runtime overhead in both development mode (Preact WMR `npm run start`) as well as production mode (static SSR with `npm run build --prerender`). Just like Twind's own "use with Preact" utility ( https://github.com/tw-in-js/use-twind-with/tree/main/packages/preact ), I implement a `Preact.options.vnode()` hook in my codebase to transform `class` (etc.) attributes, so it would not make sense to preserve `tw'...'` in this kind of code (as Twind would be called twice on the same expressions ... which would not output duplicate CSS rules thanks to predictable hashing / caching, but it would still waste runtime resources in backend pre-render and frontend hydration or render):
```jsx
<div class={tw`bg-red-400`}>
</div>
```
(PS: I implemented several micro-optimisations related to Preact's `VNode` render lifecycle in my app, thus why I am not using Twind's off-the-shelf "use with Preact" utility ... I will open-source my experimental Preact WMR setup when it is a bit more stable)

So, here's my WMR `transform` plugin:

`wmr.config.mjs`
=>
```js
const TWIND_REGEXP = /twind`([\sa-zA-Z0-9:\-()@.,/!&]+)`/gm;

export default async function (config) {
	/**
	 * @type {import('rollup').Plugin}
	 */
	const twindPlugin = {
		name: 'Twind WMR plugin',

		transform(code, id) {
			if (!/\.[tj]sx?$/.test(id)) return;
			if (id[0] === '\0' || id[0] === '\b') return;

			const func = (_match, $1) => {
				const classes = $1.replace(/\s\s+/g, ' ');
				const ret = `\`${classes}\``;
				return ret;
			};
			if (TWIND_REGEXP.test(code)) {
				code = code.replace(TWIND_REGEXP, func);
			}

			return { code, map: null };
		},
	};
	config.plugins.push(twindPlugin);
}
```

...as you can see, I use a special `twind` tag instead of `tw`, so as to not bring confusion in the codebase. I therefore need to configure the TypeScript compiler accordingly for the `@twind/typescript-plugin`:

```json
{
	"compilerOptions": {
		"plugins": [
			{
				"name": "@twind/typescript-plugin",
				"tags": ["tw", "apply", "twind"]
			}
		]
	}
}
```

...and most importantly, the `twind` tag must have an implementation (even though it is actually erased for runtime) so that it doesn't trigger errors in Visual Studio Code. This is effectively a no-op function that returns the "plain-text" string parts of the tagged template litteral (this assumes no interpolations are used, which is true in my case):

```ts
export const twind = (strings: TemplateStringsArray, ..._interpolations: string[]): string => {
	if (strings && strings.length === 1) {
		return strings[0];
	}
	throw new Error('Twind no-op interpolated string template?!');
	// let res = '';
	// const l = strings ? strings.length : 0;
	// for (let i = 0; i < l; i++) {
	// 	res += strings[i];
	// 	if (i < l - 1) {
	// 		res += interpolations[i];
	// 	}
	// }
	// return res;
};
```

**Comment by sastan** [comment]
> (PS: I implemented several micro-optimisations related to Preact's `VNode` render lifecycle in my app, thus why I am not using Twind's off-the-shelf "use with Preact" utility ... I will open-source my experimental Preact WMR setup when it is a bit more stable)

That sounds very interesting. Could you share some things you have done. Maybe in private if it is closed source?

**Comment by sastan** [comment]
> This is effectively a no-op function that returns the "plain-text" string parts of the tagged template litteral (this assumes no interpolations are used, which is true in my case):

I have been using [plain-tag](https://www.npmjs.com/package/plain-tag) for cases like this.

```js
import twind from 'plain-tag';

const className = twind`...`
```

**Comment by sastan** [comment]
> Are you able to successfully trigger the Twind VSCode plugin inside arbitrary strings?

And now to your real question... That is the goal. I hope to support all the following cases:

```js
style({
  base: `...`,
})

style({
  base: {
   '@apply': '...',
  },
})

<div tw="...">

// Only if explicitly enabled
<div class="...">
<div class ={`..`}>
<div className="...">
<div className={`..`}>

// Far in the future
`bg-${active ? 'red' : 'purple'}-500`
```

**Comment by danielweck** [comment]
My codebase is currently closed-source only because of its "rapidly evolving" nature (i.e. crappy experimental code without unit tests) :D

I am more than happy to share thoughts + relevant fragments of code.

I have been posting issues in the Preact WMR repository too, because of problematic / blocking issues with static SSR (route pre-rendering + lazy components and hydration) ... at the moment I am using layers of workarounds to ensure that my implementation covers all edge cases.

**Comment by danielweck** [comment]
> I have been using [plain-tag](https://www.npmjs.com/package/plain-tag) for cases like this.

Thanks for the tip 👍 

https://github.com/WebReflection/plain-tag/blob/master/esm/index.js
```js
export default function (t) {
  for (var s = t[0], i = 1, l = arguments.length; i < l; i++)
    s += arguments[i] + t[i];
  return s;
};
```

...though I must say I prefer using the spread / rest operator (and not import a whole NPM package for such as tiny function):

```ts
export default function (strings: TemplateStringsArray, ...interpolations: string[]): string {
	let s = strings[0];
	const l = interpolations.length;
	for (let i = 0; i < l; i++) {
		s += interpolations[i] + strings[i + 1];
	}
	return s;
}
```

:)

**Comment by danielweck** [comment]
> > Are you able to successfully trigger the Twind VSCode plugin inside arbitrary strings?
> 
> And now to your real question... That is the goal. I hope to support all the following cases:

Just a quick update to let you know that I have been testing the latest Twind versions, and only tagged template strings are supported by the plugin. Once again, this is absolutely fine by me, as I use them everywhere with a no-op implementation :) (which is "erased" at build time, so that there are no unnecessary runtime costs) I just wanted to follow-up on this issue, for those who actually wish to use the plain `class[Name]` props in their JSX, or elsewhere.

```
  "css.validate": false,
  "less.validate": false,
  "scss.validate": false,
  "editor.quickSuggestions": {
    "strings": true
  }
```

**Comment by sastan** [comment]
That should be fixed with 0.3.0.

@danielweck Could you check if all your cases are handled?

**Comment by danielweck** [comment]
Works great :)

I personally deactivate VSCode's  `"editor.quickSuggestions": { "strings": true }`, but it's good to know it works! 👍
